### PR TITLE
Rename SelfAux to ProjectionAux

### PR DESCRIPTION
**Pull request by GiudGiud**
Adds tests for projecting as well

TODO (potential) :
- [x] add warning or info message for inexact projections due to shape/orders
- [ ] add warning or info message for inexact projections due to insufficient quadrature
- [x] add error on unsupported FE families for source variable (if any)
- [x] add error on unsupported FE families for target variable
- [x] investigate SZABAB higher order 1d & 2d projection issue
- [x] investigate higher order side_hierarchic 1d & 2d projection issue
- [x] investigate monomial -> lagrange 1d projection issue
- [ ] understand all that can be wrong with this approach
- [x] 2d tests on quads
- [x] 2d tests on tris
- [x] valgrind the memory error for 2D

my current take on the errors is:
- side_hierarchic isnt supported. Setting dofs indirectly through auxkernels simply doesnt work for these variables
- szabab, lagrange and other nodal variables needed a projection from elemental to nodal
- nodal higher order was problematic because the lower order nodal source didnt have the right value at the center node (indexing at _qp=0 but _qp = 0 for an edge node)

Higher order Szabab still looks odd. Where are the dofs for those?
Screen Shot 2023-05-01 at 2 56 11 PM

### CODE REVIEW

File: test/tests/auxkernels/projection_aux/1d.i
Comment (GiudGiud) [inline test/tests/auxkernels/projection_aux/1d.i:52]: @roystgnr is any of this not true?

**Comment by GiudGiud** [comment]
so what we need here is a projection rule from elemental to nodal variables, as the current logic isnt valid, and fails silently once the right boolean is set to allow elemental variables coupled into nodal.

File: test/tests/auxkernels/projection_aux/1d.i
Comment (GiudGiud) [reply test/tests/auxkernels/projection_aux/1d.i:52]: expanded the list in current version of file

File: framework/src/auxkernels/ProjectionAux.C
Comment (GiudGiud) [inline framework/src/auxkernels/ProjectionAux.C:36]: it s absolutely crazy that it s all it took to get mpi working

File: framework/src/auxkernels/SelfAux.C
Comment (GiudGiud) [inline framework/src/auxkernels/SelfAux.C:1]: I git mved, but git still thinks i deleted

File: framework/src/auxkernels/ProjectionAux.C
Comment (lindsayad) [inline framework/src/auxkernels/ProjectionAux.C:63]: The `else` cases made me think of nodal patch recovery. @hugary1995 am I wrong for thinking of that?

Comment (hugary1995) [reply framework/src/auxkernels/ProjectionAux.C:63]: Ya, both are least squares fitting essentially. My understanding is that auxkernel doesn't construct a larger patch of element -- it's just extrapolating into the nodes, whereas NPR does the fitting over all nodal neighbors.

**Comment by lindsayad** [comment]
I'm going to defer to @roystgnr on this one and unsubscribe. If you need me, just mention me

File: framework/src/auxkernels/ProjectionAux.C
Comment (GiudGiud) [inline framework/src/auxkernels/ProjectionAux.C:87]: going to have to add something here when the variable isnt defined on that element

Comment (GiudGiud) [reply framework/src/auxkernels/ProjectionAux.C]: done

**Comment by GiudGiud** [comment]
Passing now

**Comment by GiudGiud** [comment]
so at this point I believe this code in AuxKernel is not fit for **setting** variables that have DOFs on elements and nodes
```
  if (isNodal()) /* nodal variables */
  {
    if (_var.isNodalDefined())
    {
      _qp = 0;
      ComputeValueType value = computeValue();
      // update variable data, which is referenced by other kernels, so the value is up-to-date
      _var.setNodalValue(value);
    }
  }
```

isNodal() ends up relying on:
```
  _is_nodal = (_continuity == C_ZERO || _continuity == C_ONE);
```
in MooseVariableData.C
which imo is pretty ad-hoc.

I think isNodal could become either, to be clearer,
- isOnlyNodal
- hasOnlyDoFsOnNodes

In the meantime, I would like for this PR to go under review as it's needed to support an ANL project

thoughts?

**Review by roystgnr — CHANGES_REQUESTED** [review]
If you're in a hurry we can just add a mooseError for all the non-supported target variable types ... but I really want to at least figure out what's going wrong in the p=0 `MONOMIAL` to p=1 `LAGRANGE` case before we merge this.

File: test/tests/auxkernels/projection_aux/gold/1d_linear.csv
Comment (roystgnr) [reply test/tests/auxkernels/projection_aux/gold/1d_linear.csv:2]: Never mind; I think I get it now.  We're evaluating `ElementL2Difference` here, not from the exact IC, but from the `base_elem` approximation to the exact IC.  Even with an exact Lagrange solution we'd still be seeing the error in `base_elem`.

Comment (roystgnr) [reply test/tests/auxkernels/projection_aux/gold/1d_linear.csv:2]: No, no, that's still not the case.  You're bumping up `AuxVariables/base_elem/order=FIRST` in `tests`/`1d_linear`.  That should still recover the exact solution.

File: framework/src/auxkernels/ProjectionAux.C
Comment (GiudGiud) [reply framework/src/auxkernels/ProjectionAux.C:32]: changed to algebraic. I thought it made sense to have geometric when you need element info and algebraic as well for variable info

File: framework/src/auxkernels/ProjectionAux.C
Comment (GiudGiud) [reply framework/src/auxkernels/ProjectionAux.C:79]: yup

File: framework/src/auxkernels/ProjectionAux.C
Comment (GiudGiud) [reply framework/src/auxkernels/ProjectionAux.C:87]: Umm you're right.
I added the second option last, and did not think to move that case there.

In my mind it was fair to weight the centroid value with the volume. But for using the (discontinuous) nodal value, I would have to weight with either something like the local solid angle of the element or some distance to the centroid instead of the volume
I changed it. I just want a projection rule, it will have to be approximate anyway

File: test/tests/auxkernels/projection_aux/gold/1d_linear.csv
Comment (GiudGiud) [reply test/tests/auxkernels/projection_aux/gold/1d_linear.csv:2]: >  but we have 0.0258 error when projecting to a linear LAGRANGE?

that was from using the centroid value near a boundary. It was basically forcing a flat profile near the boundary

> we can chalk up p=2 SZABAB to our previous WTF ...

all nodal variables unsupported by the auxkernel system are disallowed now

File: test/tests/auxkernels/projection_aux/gold/1d_constant.csv
Comment (GiudGiud) [reply test/tests/auxkernels/projection_aux/gold/1d_constant.csv:2]: > Any idea why the p=2 SZABAB cases are failing here?

AuxKernels only set the nodal value. None of the additional dofs. 
Same for all the others

**Comment by GiudGiud** [comment]
not ready yet, trying to change something in THM now

**Comment by GiudGiud** [comment]
ok done now! Turns out it was an easy fix with @joshuahansel 's insights

Ready for re-review
@joshuahansel

File: framework/src/auxkernels/ProjectionAux.C
Comment (roystgnr) [reply framework/src/auxkernels/ProjectionAux.C]: It does make sense!  So much sense that we don't ever do the latter without the former.

**Comment by GiudGiud** [comment]
Of course Griffin uses both the SelfAux and their own CopyAux...

**Comment by GiudGiud** [comment]
Green now

**Review by roystgnr — APPROVED** [review]
I'm not seeing any more problems (or at least nothing that I'd call a bug rather than a hard-to-add missing feature...)

**Comment by GiudGiud** [comment]
Awesome! Thanks a lot for the review! This has bugged me for a long time now.

File: framework/src/auxkernels/ProjectionAux.C
Comment (lindsayad) [reply framework/src/auxkernels/ProjectionAux.C]: Maybe I shouldn't have unsubscribed. We don't add MOOSE algebraic ghosting functors to the mesh, so the `|` is needed

File: test/tests/auxkernels/projection_aux/1d.i
Comment (lindsayad) [reply test/tests/auxkernels/projection_aux/1d.i:50]: Probably not

File: framework/src/auxkernels/ProjectionAux.C
Comment (lindsayad) [reply framework/src/auxkernels/ProjectionAux.C]: So this works because of the default geometric ghosting in libMesh (and also because of Exodus output turned on in this input file). If I disable the default geometric ghosting in libMesh (which we never do in MOOSE so this is somewhat of a hypothetical exercise) and turn off exodus output, then I get some exodiffs with distributed mesh

Comment (lindsayad) [reply framework/src/auxkernels/ProjectionAux.C]: I don't understand why the heck I added `MooseGhostPointNeighborLayers` when we have `ElementPointNeighborLayers`. Opened https://github.com/idaholab/moose/issues/24573 to remove it

Comment (roystgnr) [reply framework/src/auxkernels/ProjectionAux.C]: >We don't add MOOSE algebraic ghosting functors to the mesh

And ... of course we don't.  `/*to_mesh = */ false`.  I'd forgotten that I'd given up on fighting the `break_encapsulation_because_thats_easier_to_optimize` option; sorry for both stages of that.

**Comment by GiudGiud** [comment]
@yukkiezy and @aprilnovak

**Pull request merged**
